chichiwang/exercise-js
# Build a CSS version of the Logo

## Description
Linked to Issue: #81 
Create a CSS version of the logo and use it instead of `static/images/logo.webp`. There are two reasons for this:
1. A CSS or inline-SVG logo is both more lightweight and loads faster than a referenced `img` tag
2. We need a programmatic version of the logo to open up the options of using CSS to animate splash images, or to add transition effects on user events like mouse overs

The intent is for the logo to be entirely CSS (with the exception of favicons, which are restricted to external image assets). In the future perhaps we can experiment with animations using the logo as a basis of it.

This PR introduces changes across a large number of files:
* All page snapshots use the `Layout` component, which renders the `Nav` which contains the icon
* The no-longer used `static/images/logo.webp` image was removed from the repository
* Helpers were created to allow the CSS logo to be both resizeable while enforcing its original aspect ratio

## Changes
* `.eslintrc` custom rules updated:
  * [react/jsx-props-no-spreading](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-props-no-spreading.md) disabled: when determining a large set of conditional props to feed a React component, it is my opinion that having a separate variable per prop (including those which will be `undefined`) hampers legibility more than a props object
  *  [react/require-default-props](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/require-default-props.md) is disabled : in the case an optional prop defaults to `undefined`, it is uglier and more cumbersome to set a default value than not
  * [@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars/) rule added to the typescript file overrides: eliminates an erroneous lint failure when [declaring a function type definition](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-type-expressions)
* `app/var.module.scss`: Hex value for `white` added to `$colors` map
* `pages/Home/index.tsx`: Replace `<img>` tag for `logo.webp` with the new `<Logo>` component
* `pages/Layout/Nav.tsx`: Replace `<img>` tag for `logo.webp` with the new `<Logo>` component
* `pages/components/Logo/Egret.tsx`: Resizeable inline-SVG of the egret silhouette in the logo
* `pages/components/Logo/getDimensions.ts`: utility for enforcing dimensions props, and for providing missing dimension values that maintain the aspect ratio for the logo
* `pages/components/Logo/index.tsx`: Resizeable, styleable CSS logo component
* Remove file `static/images/logo.webp`: no longer used

## Steps to QA
* Pull down and switch to this branch
* Run the server and verify the logo in the browser
  * The `/` homepage route and the logo in the nav should look identical to the `master` branch
